### DISCUSSION TITLE
Best practice for soft-delete with unique constraint

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#10152] [Usage Questions]
I have a bunch of hacks I have been doing for this case. I am curious to understand if there is a recommended best practice that sqlalchemy suggests using as I believe it is common.
PS: I have gone through previous issues and docs but I may have missed any recommendations here.

I have a case where I can have rows that are soft deleted.
Think of a table like a `users` table with `user_name` being a unique value:

user_name | user_id | ... | status
-|-|-|-
userA | 1 | ... | Active
userB | 2 | ... | Deactivated
userC | 3 | ... | Dormant

Here, I had 3 users. And `userB` had previously decided to deactivate their account.
Now, I want a new user to be able to come in and register with the `user_name=userB` because that is now available and not actively being used.

Solutions I have considered so far:
1. Don't have a unique constraint check. And just do an `IF` condition in my application logic
    - Con: Need to handle race conditions. And handling race conditions across DBs is nontrivial? For example, sqlite does not support `with_for_update()`
2. Don't allow users to re-use usernames
    - Con: Causes clutter over time
3. Create a separate `user_deactivated` table where the Deactivated users are "moved" when they want to deactivate
    - Con: In my real use case I need to query the deactivated records frequently. And then unioning 2 tables etc gets messy every time I need to read it. And again some DBs have differences in what is allowed/not-allowed in `UNIONed` queries making some things like groupby or pagination very inefficient/throw errors (I'm looking at you - MySQL)

**Comment by CaselIT** [comment]
Hi,

On postgresql you can define a where condition on an index, so you could add an unique index on user name where status is not dormant.

For a more general approach maybe appending a random string to the end of the dorman usernames could maybe be an option? Or having a second column that includes a random string/number (or even sequential) that's only populated on dormant users, then the unique constraint would be on the two columns (and you can even add a check contraint to ensure that this second column is a fixed value if status is not dormant)

**Reply by AbdealiLoKo** [reply]
I see.
Just curious - another thing I came across is a CHECK constraint. Do you think a custom check constraint would help here ?

**Reply by AbdealiLoKo** [reply]
Also, if I decide to do this check in my application layer - i.e. in python before I do a `add()` / `commit()` 
Is there any recommendation on whether locking can ever achieve the need here  ?

**Reply by CaselIT** [reply]
> Do you think a custom check constraint would help here ?

I don't think so, since they have access only to a single row at a time

> Is there any recommendation on whether locking can ever achieve the need here ?


Probably a for update lock is required

**Reply by AbdealiLoKo** [reply]
The `FOR UPDATE` does not seem to help me here - cause it seems to lock the flush.

So, if I have:
```
# When trying to create a new `item`
data = User.query.filter_by(name=item.name).with_for_update().all()
if len(data) > 0:
    raise Exception(...)
time.sleep(5)  # do some other logic here which is slow (to test race conditions)
session.flush(item)
```

The `data` is read by both the python shells I am using to test this case.
The `session.flush(item)` is the line that blocks and does not run until the other shell completes it's commit.

What I wanted was that `data = User.query.filter_by(name=item.name).with_for_update().all()` <- this line does not run and waits for the other shell to finish.

That seems to be a `EXCLUSIVE LOCK` as compared to a `FOR UPDATE` lock ?
Ref: https://stackoverflow.com/questions/37331646/sqlalchemy-explicit-locking-of-postgresql-table

Are there any helpers in sqlalchemy for that or I just call the DB commands directly ?

**Reply by CaselIT** [reply]
not directly in sqlalchemy. You can either execute the string directly, create a custom object and use this extension to compile it https://docs.sqlalchemy.org/en/20/core/compiler.html#module-sqlalchemy.ext.compiler or look if there is something in 3-rd party libs, like on sqlalchemy-utils
